### PR TITLE
feature/hosts

### PR DESCRIPTION
**Pull request by ryankurte**
hey thanks for making a neat (and well timed) thing!

this PR adds hosts module with hooks and macros, definitely needs a review for correctness.

### CODE REVIEW

File: libnss/src/host.rs [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+                        libc::memcpy(p.as_ptr() as *mut libc::c_void, addr as *mut libc::c_void, 16);
+                        IpAddr::V6(Ipv6Addr::from(p))
+                    },
+                    _ => {
+                        //error!("address length and format mismatch (length: {}, format: {})", len, format);
+                        return NssStatus::NotFound.to_c();
+                    }
+                };
+
+
```
Comment (csnewman) [inline libnss/src/host.rs]: Formatting

File: libnss/src/host.rs [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+                }
+            }
+        }
+
+        // Write null termination
+        libc::memset(array_pos as *mut libc::c_void, 0, 1);
+
+        // Set single / first address
+        (*hostent).h_addr = *(*hostent).h_addr_list;
+
```
Comment (csnewman) [inline libnss/src/host.rs]: Formatting

File: libnss/src/host.rs [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+            Addresses::V6(addrs) => {
+                for a in addrs {
+                    let ptr = buffer.reserve(addr_len);
+
+                    let o = a.octets();
+                    libc::memcpy(ptr as *mut libc::c_void, o.as_ptr() as *mut libc::c_void, addr_len as usize);
+                    
+                    *array_pos = ptr;
+
+                    array_pos = array_pos.offset(1);
```
Comment (csnewman) [inline libnss/src/host.rs:71]: Seems quite spaced out

File: libnss/src/host.rs [hunk 4]
```diff
@@ -0,0 +1,233 @@
+
+use crate::interop::CBuffer;
+use std::mem;
+use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};
+use std::collections::VecDeque;
+
+
```
Comment (csnewman) [inline libnss/src/host.rs:5]: Formatting

**Review by csnewman — CHANGES_REQUESTED** [review]

**Comment by csnewman** [comment]
Generally great, just some minor style choices.

**Comment by ryankurte** [comment]
i gave it a `cargo fmt` pass and removed a little bit of whitespace, but, i'm not sure what your other formatting comments refer to..? 

(also i only applied the new formatting to this file to avoid burying the PR in noise, you might consider running it against the whole repo too)

File: libnss/src/host.rs (hunk 1 above)
Comment (csnewman) [reply libnss/src/host.rs]: Should only be one new whiteline here

File: libnss/src/host.rs [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+                            val.to_c_hostent(hostbuf, &mut buffer);
+                            NssStatus::Success.to_c()
+                        },
+                        None => NssStatus::NotFound.to_c()
+                    },
+                    Err(_) => NssStatus::NotFound.to_c()
+                }
+            }
+
```
Comment (csnewman) [inline libnss/src/host.rs:229]: No need for 3 whitelines

**Review by csnewman — CHANGES_REQUESTED** [review]

**Comment by csnewman** [comment]
Thanks

File: libnss/src/host.rs [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+}
+
+impl Host {
+    pub unsafe fn to_c_hostent(self, hostent: *mut CHost, buffer: &mut CBuffer) {
+        (*hostent).name = buffer.write_str(self.name);
+        (*hostent).h_aliases = buffer.write_strs(&self.aliases);
+
+        let ptr_size = mem::size_of::<*mut libc::c_char>() as isize;
+
+        let (addr_len, count) = match &self.addresses {
```
Comment (bjeanes) [inline libnss/src/host.rs:24]: If I am understanding this correctly, this makes it possible to only return a list of V4 addresses OR a list of V6 addresses, but not possible to return a list of all V4 AND V6 addresses for a host.

Assuming my interpretation is correct, is this an inherit limitation of the NSS C API or just of this shim?

Comment (bjeanes) [reply libnss/src/host.rs:24]: I read the https://ftp.gnu.org/old-gnu/Manuals/glibc-2.2.3/html_chapter/libc_16.html#SEC318 linked in a comment in this diff and it does appear to be inherent in the shape of the struct we need to return. Bummer.

**Comment by bjeanes** [comment]
I might have made an error that I've missed, but `get_host_by_name` does not appear to be called. I have added logger statements to the three functions and when using `getent hosts` I see that `get_all_entries` is called. `getent hosts 1.2.3.4` shows that `get_host_by_addr` is called, but `getent hosts foo.bar` indicates that `get_host_by_name` is NOT called.

Furthermore, if I return `None` (or `vec![]` for `get_all_entries`) all is fine, but if I actually return a `Host`, I get a segmentation fault! I think the host struct may not be getting laid out correctly?

**Comment by bjeanes** [comment]
> `getent hosts foo.bar` indicates that `get_host_by_name` is NOT called.

On my system, at least, the function is expected to be called the equivalent of `[<_nss_ $mod_ident _gethostbyname2_r>]` (note the trailing `2`).

I have looked at some other source and mailing lists and it seems like this is all a mess. At the very least, having the non-`2` version just call the first one seems like a viable patch. I'll get a bit further with my changes and then open a PR that brings it closer to what seems to be done in practice: https://github.com/lathiat/nss-mdns/blob/58fcdd2071eb64372f071cee14982536fd4c7a99/src/nss.c#L215-L262

See also https://libc-alpha.sourceware.narkive.com/Bouz4PWh/patch-bz-16077-get-canonical-name-in-getaddrinfo-from-hosts-file-for-af-inet

**Comment by bjeanes** [comment]
> if I actually return a `Host`, I get a segmentation fault!

I am still not sure about these, so I am going to go over the `CHost` construction carefully and see if I can spot anything obvious...

File: libnss/src/host.rs [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+    fn get_host_by_addr(addr: IpAddr) -> Option<Host>;
+}
+
+/// NSS C Host object
+/// https://ftp.gnu.org/old-gnu/Manuals/glibc-2.2.3/html_chapter/libc_16.html#SEC318
+#[repr(C)]
+#[allow(missing_copy_implementations)]
+pub struct CHost {
+    pub name: *mut libc::c_char,
+    pub h_aliases: *mut libc::c_char,
```
Comment (bjeanes) [inline libnss/src/host.rs:98]: I think this should be:

``` suggestion
    pub h_aliases: *mut *mut libc::c_char,
```

I think this is the cause of my segfaults. https://ftp.gnu.org/old-gnu/Manuals/glibc-2.2.3/html_chapter/libc_16.html#SEC318 defines its type as `char **h_aliases`. Changing it in my local copy fixes the segfault but I must have a different error with my pointer fiddling as `getent` prints out the first alias but then gibberish for the second one (but doesn't segfault!).

**Pull request merged**